NamedExports from barrel files are not dynamically imported for lazy loading properly.

# 🐞 Bug report

### Command (mark with an `x`)
```
- [x] build
- [x] serve
```

### Is this a regression?

<!-- Did this behavior use to work in the previous version? -->
<!-- ✍️--> Yes, the previous version in which this bug was not present was: ....
Yes, sort of. This worked in 7 with the loadChildren strings but does not work in 8.

### Description

<!-- ✍️--> A clear and concise description of the problem...

This was originally reported in https://github.com/nrwl/nx/issues/1471.

This is an extension of #14399. The fix for that issue solves it for unnamed ExportDeclarations but not for NamedExports. 

```ts
// export * from './lib/my.module'; // This works
export { MyModule } from './lib/my.module'; // This does not work
```

## 🔬 Minimal Reproduction
https://github.com/FrozenPandaz/demos/blob/lazybug/libs/lib1/src/index.ts

## 🔥 Exception or Error
```
ERROR in ./src/app/app.module.ts
Module not found: Error: Can't resolve './../../../../libs/lib1/src/index.ngfactory' in '/Users/jason/projects/temp/temp/nx8-test/apps/app1/src/app'
```

## 🌍 Your Environment
`8.0.1`

**Anything else relevant?**
May affect Nx Users.

cc @filipesilva who resolved #14399

--- a/packages/ngtools/webpack/src/transformers/import_factory.ts
+++ b/packages/ngtools/webpack/src/transformers/import_factory.ts
@@ -10,6 +10,9 @@ import * as ts from 'typescript';
 import { forwardSlashPath } from '../utils';
 
 
+// Check if a ts.Symbol is an alias.
+const isAlias = (symbol: ts.Symbol) => symbol.flags & ts.SymbolFlags.Alias;
+
 /**
  * Given this original source code:
  *
@@ -193,15 +196,26 @@ function replaceImport(
 
   // Try to resolve the import. It might be a reexport from somewhere and the ngfactory will only
   // be present next to the original module.
-  const exportedSymbol = typeChecker.getSymbolAtLocation(exportNameId);
+  let exportedSymbol = typeChecker.getSymbolAtLocation(exportNameId);
   if (!exportedSymbol) {
     return warnAndBail();
   }
+  // Named exports are also a declaration in the re-exporting module so we have to follow the
+  // re-exports to find the original symbol.
+  if (isAlias(exportedSymbol)) {
+    exportedSymbol = typeChecker.getAliasedSymbol(exportedSymbol);
+    if (!exportedSymbol) {
+      return warnAndBail();
+    }
+  }
+
+  // Find declarations of the original symbol so we can get their source file name.
   const exportedSymbolDecl = exportedSymbol.getDeclarations();
   if (!exportedSymbolDecl || exportedSymbolDecl.length === 0) {
     return warnAndBail();
   }
 
+  // Let's guess the first declaration is the one we want, because we don't have a better criteria.
   // Get the relative path from the containing module to the imported module.
   const relativePath = relative(dirname(fileName), exportedSymbolDecl[0].getSourceFile().fileName);
 

--- a/packages/ngtools/webpack/src/transformers/import_factory_spec.ts
+++ b/packages/ngtools/webpack/src/transformers/import_factory_spec.ts
@@ -64,9 +64,83 @@ describe('@ngtools/webpack transformers', () => {
       expect(warningCalled).toBeTruthy();
     });
 
-    it('should support resolving reexports', () => {
+    it('should support resolving * re-exports', () => {
       const additionalFiles: Record<string, string> = {
         'shared/index.ts': `
+          export * from './path/to/lazy/lazy.module';
+        `,
+        'shared/path/to/lazy/lazy.module.ts': `
+          export const LazyModule = {};
+        `,
+      };
+      const input = tags.stripIndent`
+        const ɵ0 = () => import('./shared').then(m => m.LazyModule);
+        const routes = [{
+          path: 'lazy',
+          loadChildren: ɵ0
+        }];
+      `;
+
+      const output = tags.stripIndent`
+        const ɵ0 = () => import("./shared/path/to/lazy/lazy.module.ngfactory").then(m => m.LazyModuleNgFactory);
+        const routes = [{
+          path: 'lazy',
+          loadChildren: ɵ0
+        }];
+      `;
+
+      const { program, compilerHost } = createTypescriptContext(input, additionalFiles, true);
+      const transformer = importFactory(() => { }, () => program.getTypeChecker());
+      const result = transformTypescript(undefined, [transformer], program, compilerHost);
+
+      expect(tags.oneLine`${result}`).toEqual(tags.oneLine`${output}`);
+    });
+
+    it('should support resolving named re-exports', () => {
+      const additionalFiles: Record<string, string> = {
+        'shared/index.ts': `
+          export { LazyModule } from './path/to/lazy/lazy.module';
+        `,
+        'shared/path/to/lazy/lazy.module.ts': `
+          export const LazyModule = {};
+        `,
+      };
+      const input = tags.stripIndent`
+        const ɵ0 = () => import('./shared').then(m => m.LazyModule);
+        const routes = [{
+          path: 'lazy',
+          loadChildren: ɵ0
+        }];
+      `;
+
+      const output = tags.stripIndent`
+        const ɵ0 = () => import("./shared/path/to/lazy/lazy.module.ngfactory").then(m => m.LazyModuleNgFactory);
+        const routes = [{
+          path: 'lazy',
+          loadChildren: ɵ0
+        }];
+      `;
+
+      const { program, compilerHost } = createTypescriptContext(input, additionalFiles, true);
+      const transformer = importFactory(() => { }, () => program.getTypeChecker());
+      const result = transformTypescript(undefined, [transformer], program, compilerHost);
+
+      expect(tags.oneLine`${result}`).toEqual(tags.oneLine`${output}`);
+    });
+
+
+    it('should support resolving re-export chains', () => {
+      const additionalFiles: Record<string, string> = {
+        'shared/index.ts': `
+          export { LazyModule } from './index2';
+        `,
+        'shared/index2.ts': `
+          export * from './index3';
+        `,
+        'shared/index3.ts': `
+          export { LazyModule } from './index4';
+        `,
+        'shared/index4.ts': `
           export * from './path/to/lazy/lazy.module';
         `,
         'shared/path/to/lazy/lazy.module.ts': `

